source/MaterialXGenShader/ShaderGraph.cpp

```diff
@@ -481,6 +486,109 @@ ShaderGraphPtr ShaderGraph::create(const ShaderGraph* parent, const string& name
         // Start traversal from this output
         root = output;
     }
+    else if (element->isA<Node>())
+    {
+        NodePtr node = element->asA<Node>();
+
+        NodeDefPtr nodeDef = node->getNodeDef();
+        if (!nodeDef)
+        {
+            throw ExceptionShaderGenError("Could not find a nodedef for node '" + node->getName() + "'");
+        }
+
+        graph = std::make_shared<ShaderGraph>(parent, name, element->getDocument(), context.getReservedWords());
+
+        // Create input sockets
+        graph->addInputSockets(*nodeDef, context);
+
+        // Create output sockets
+        graph->addOutputSockets(*nodeDef);
+
+        // Create this shader node in the graph.
+        const string& newNodeName = node->getName();
+        ShaderNodePtr newNode = ShaderNode::create(graph.get(), newNodeName, *nodeDef, context);
+        graph->_nodeMap[newNodeName] = newNode;
+        graph->_nodeOrder.push_back(newNode.get());
+
+        // Connect it to the graph output
+        //
+        // TODO: Support multioutput nodes
+        //
+        ShaderGraphOutputSocket* outputSocket = graph->getOutputSocket();
+        outputSocket->makeConnection(newNode->getOutput());
+        outputSocket->setPath(node->getNamePath());
+
+        // Handle node parameters
+        for (ParameterPtr elem : nodeDef->getActiveParameters())
+        {
+            ShaderGraphInputSocket* inputSocket = graph->getInputSocket(elem->getName());
+            ShaderInput* input = newNode->getInput(elem->getName());
+            if (!inputSocket || !input)
+            {
+                throw ExceptionShaderGenError("Node parameter '" + elem->getName() + "' doesn't match an existing input on graph '" + graph->getName() + "'");
+            }
+
+            ParameterPtr paramPtr = node->getParameter(elem->getName());
+            if (paramPtr)
+            {
+                ValuePtr value = paramPtr->getResolvedValue();
+                if (value)
+                {
+                    inputSocket->setValue(value);
+                }
+
+                inputSocket->setPath(paramPtr->getNamePath());
+                input->setPath(inputSocket->getPath());
+
+                const string& unit = paramPtr->getUnit();
+                if (!unit.empty())
+                {
+                    inputSocket->setUnit(unit);
+                    input->setUnit(unit);
+                }
+            }
+
+            // Connect to the graph input
+            inputSocket->makeConnection(input);
+        }
+
+        // Handle node inputs
+        for (const InputPtr& elem : nodeDef->getActiveInputs())
```

Comment:
Wondering if the logic in these 2 loops can be reused or a getActiveInterfaceElements() should be added as it looks like they do pretty well the same thing to update sockets.

Reply:
Good point. And there is already a getActiveValueElements which I can use here.